### DISCUSSION TITLE
Emacs

### DISCUSSION THREAD

**Discussion by jsbien** [psb1558/Junicode-font#47] [Show and tell]
I just learned that you can set font in command line, e.g.
`emacs -fn "DejaVu Sans Mono-12"`
(an example from https://www.gnu.org/software/emacs/manual/html_node/emacs/Font-X.html#Font-X).
So I tried
`emacs -fn "JuniusX"`
and it worked. It's not exactly what I want, but I asked to close my bug report.
Actually I tried first
`emacs -fn "JuniusX-subset"`
`emacs -fn "JuniusX-Regular-subset"`
because I'm quite confused what the actual font name is.
![EmacsPUA 2020-12-12 19-31-40](https://user-images.githubusercontent.com/8024702/101992132-45089380-3cb1-11eb-8d37-33ec03a7f7dd.png)

**Comment by emacsomancer** [comment]
You can also dynamically set the font (or set it in the `init.el` configuration file), e.g.
`(set-face-attribute 'default nil :family "JuniusX")` , or, probably better, just set it for the variable-pitch proportional face, e.g. `(set-face-attribute 'variable-pitch nil :family "JuniusX")` [see [https://gitlab.com/jabranham/mixed-pitch](https://gitlab.com/jabranham/mixed-pitch) for mixing proportional and fixed faces in the same buffer]

**Reply by jsbien** [reply]
Thanks for your suggestions, but actually I don't want it to be the default font. This is rather a long story, cf. https://debbugs.gnu.org/cgi/bugreport.cgi?bug=45184.

**Reply by emacsomancer** [reply]
I'm not sure I followed all of it. Are you trying to use JuniusX (or some font) to supplement missing characters from whatever your default font is?

**Comment by jsbien** [comment]
Yes, this is a good summary. Such supplementation worked for me for years for MUFI PUA characters and I want to keep it this way. However it  doesn't work when a PUA character is from a private plane.

**Reply by emacsomancer** [reply]
Were you able to get something satisfactory working with `fontset-default`, which I think is what Eli ends up suggesting?

E..g., I have in my `init.el`:

````
;; set devanagari & other fonts
(defun bms/font-spec ()
  "Set typefaces for certain Unicode ranges."
  (interactive)
                                        ; fallback font?
  (set-fontset-font "fontset-default" 'nil (font-spec :name "DejaVu Sans Mono")) ; general default
  (set-fontset-font "fontset-default" '(#x16A0 . #x16F8) (font-spec :size 18 :name "Noto Sans Runic")) ; runic
  (set-fontset-font "fontset-default" '(#x1680 . #x169F) (font-spec :name "Noto Sans Ogham")) ; ogham
  (set-fontset-font "fontset-default" '(#x10330 . #x1034A) (font-spec :size 20 :name "Junicode")) ; gothic
  (set-fontset-font "fontset-default" '(#x11000 . #x1107F) (font-spec :name "Noto Sans Brahmi")) ; brahmi  
  (set-fontset-font "fontset-default" '(#x0980 . #x09FF) (font-spec :name "Noto Sans Bengali")) ; bengali
  (set-fontset-font "fontset-default" '(#x0A80 . #x0AFF) (font-spec :name "Noto Sans Gujarati")) ; gujarati
  (set-fontset-font "fontset-default" '(#x0A00 . #x0A7F) (font-spec :name "Noto Sans Gurmukhi")) ; gurmukhi  
  (set-fontset-font "fontset-default" '(#x0B00 . #x0B7F) (font-spec :name "Noto Sans Oriya")) ; oriya  
  (set-fontset-font "fontset-default" '(#x0D80 . #x0DFF) (font-spec :name "Noto Sans Sinhala")) ; sinhala
  (set-fontset-font "fontset-default" '(#x0B80 . #x0BFF) (font-spec :name "Noto Sans Tamil")) ; tamil
  (set-fontset-font "fontset-default" '(#x0C00 . #x0C7F) (font-spec :name "Noto Sans Telugu")) ; telugu
  (set-fontset-font "fontset-default" '(#x0C80 . #x0CFF) (font-spec :name "Noto Sans Kannada")) ; kannada  
  (set-fontset-font "fontset-default" '(#x0D00 . #x0D7F) (font-spec :name "Noto Sans Malayalam")) ; malayalam
  (set-fontset-font "fontset-default" '(#x0900 . #x097f) (font-spec :name "Noto Sans Devanagari SemiBold")) ; devanagari
  (set-fontset-font "fontset-default" '(#xA8E0 . #xA8FF) (font-spec :name "Noto Sans Devanagari")) ; devanagari extended 1
  (set-fontset-font "fontset-default" '(#x1CD0 . #x1CFF) (font-spec :name "Noto Sans Devanagari")) ; vedic extensions
  (set-fontset-font "fontset-default" '(#x10B00 . #x10B3F) (font-spec :name "Noto Sans Avestan")) ; avestan
  (set-fontset-font "fontset-default" '(#x10B40 . #x10B5F) (font-spec :name "Noto Sans Inscriptional Parthian")) ; inscriptional parthian   
  (set-fontset-font "fontset-default" '(#x10B60 . #x10B7F) (font-spec :name "Noto Sans Inscriptional Pahlavi")) ; inscriptional pahlavi
  (set-fontset-font "fontset-default" '(#x10B80 . #x10BAF) (font-spec :name "Noto Sans Psalter Pahlavi")) ; psalter pahlavi
  (set-fontset-font "fontset-default" '(#x103A0 . #x103DF) (font-spec :name "Noto Sans Old Persian")) ; old persian  
  (set-fontset-font "fontset-default" '(#x1200 . #x137F) (font-spec :name "Noto Sans Ethiopic")) ; ethiopic
  (set-fontset-font "fontset-default" '(#x1380 . #x139F) (font-spec :name "Noto Sans Ethiopic")) ; ethiopic supplement
  (set-fontset-font "fontset-default" '(#x2D80 . #x2DDF) (font-spec :name "Noto Sans Ethiopic")) ; ethiopic extended
  (set-fontset-font "fontset-default" '(#x2C80 . #x2CFF) (font-spec :name "Noto Sans Coptic")) ; coptic
  (set-fontset-font "fontset-default" '(#x13000 . #x1343F) (font-spec :name "Noto Sans Egyptian Hieroglyphs")) ; hieroglyphic
  (set-fontset-font "fontset-default" '(#x14400 . #x1467F) (font-spec :name "Noto Sans Anatolian Hieroglyphs")) ; anatolian hieroglyphs
  (set-fontset-font "fontset-default" '(#x12000 . #x1254F) (font-spec :name "Noto Sans Cuneiform")) ; cuneiform
  (set-fontset-font "fontset-default" '(#x10600 . #x1077F) (font-spec :name "Noto Sans Linear A")) ; linear a
  (set-fontset-font "fontset-default" '(#x10000 . #x100FF) (font-spec :name "Noto Sans Linear B")) ; linear b  
  (set-fontset-font "fontset-default" '(#x10400 . #x1044F) (font-spec :name "Noto Sans Deseret")) ; deseret
  (set-fontset-font "fontset-default" '(#x0530 . #x058F) (font-spec :name "Noto Sans Armenian")) ; armenian  
  (set-fontset-font "fontset-default" '(#x2700 . #x27BF)  (font-spec :name "Symbola"))
  (set-fontset-font "fontset-default" '(#xE270 . #xF01E) (font-spec :name "Junicode")) ; fleuron block
  ;; (set-fontset-font "fontset-default" '(#x2766 . #x2767) (font-spec :name "Symbola")) ; two fleurons
  )
  
(add-hook 'before-make-frame-hook #'bms/font-spec)
````

Or does this also have problems with private planes?

**Reply by jsbien** [reply]
Thanks for the constructive suggestion. I haven't used fontset-default for years because it appeared unnecessary for my purposes. Just made a quick test and it works. _Update: there are still some problems which I investigate._
There is still an open question: why MUFI PUA characters are displayed correctly (using Junicode) without any additional configuration, and plane15 PUA are not?

**Comment by psb1558** [comment]
This is very good to know.

**Comment by psb1558** [comment]
In Aquamacs you can do Options --> Appearance --> Font for Text-Mode, then choose JuniusX (the font changes before you dismiss the dialog box, so you can easily get the size right).

And that also works!

**Reply by jsbien** [reply]
Contrary to what I wrote in my bug report, selecting JuniusX with `S-down-mouse-1’  is working now.

**Comment by jsbien** [comment]
The present situation is that Emacs displays the characters correctly when the font is selected explicitely e.g. with set-fontset-font. I would like it to do it automatically as it is done for BMP. In due time I will consult the Emacs mailing lists and report the outcome.
FYI, I found quite convenient evaluating set-fontset-font for the buffer where I really need it by putting it into the local variables block:
`%%% eval: (set-fontset-font "fontset-default" '(#xF0000 . #xF001F) (font-spec :size 18 :name "JuniusX"))`

**Comment by jsbien** [comment]
As I mentioned in another thread, I had a problem with the Scotland flag emoji, the problem could be relevant for possible use of tags for Junicode. I hope to solve the problem with the help of https://emacs.stackexchange.com/questions/62219/how-do-i-get-colour-emoji-to-display-in-emacs. I will test it soon but not immediately.
UPDATE: it works out of the box in Emacs 28, so I don't try to understand the solution proposed in the linked post.
